crypto-common/src/generate.rs

```diff
@@ -0,0 +1,78 @@
+use core::fmt;
+use hybrid_array::{Array, ArraySize};
+
+/// Secure random generation using the system's ambient RNG.
+pub trait Generate: Sized {
+    /// Randomly generate a value of this type using the system's ambient cryptographically secure
+    /// random number generator.
+    ///
+    /// # Panics
+    /// This method will panic in the event the system's ambient RNG experiences an internal
+    /// failure.
+    ///
+    /// This shouldn't happen on most modern operating systems.
+    fn generate() -> Self {
+        Self::try_generate().expect("RNG failure")
+    }
+
+    /// Randomly generate a value of this type using the system's ambient cryptographically secure
+    /// random number generator.
+    ///
+    /// # Errors
+    /// Returns [`RngError`] in the event the system's ambient RNG experiences an internal failure.
+    fn try_generate() -> Result<Self, RngError>;
+}
+
+/// Errors occurring internally within the system's ambient cryptographically secure RNG.
+// TODO(tarcieri): switch to `getrandom::Error` when it impls `core::error::Error`?
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub struct RngError(getrandom::Error);
```

Comment:
Note: I opted to use a wrapper type for the error here so we could get a consistent `core::error::Error` impl on it.

I think it would be fine to use `getrandom::Error` directly otherwise, though whether we expose `getrandom` directly in the public API is debatable.

I think `getrandom` upgrades should probably still be considered breaking as it relies on ambient global coupling for `no_std` platforms, so if we upgrade `getrandom`, the consumer initializing the system RNG also needs to upgrade it too.

Reply:
Switched to `getrandom::Error`, re-exported as `crypto_common::RngError`